python/ray/data/tests/test_formats.py

```diff
@@ -162,6 +166,33 @@ def test_write_datasource(ray_start_regular_shared):
     assert ray.get(output.data_sink.get_rows_written.remote()) == 10
 
 
+def test_open_file_with_retry(ray_start_regular_shared):
+    class Counter:
+        def __init__(self):
+            self.retry_attempts = 0
+
+        def foo(self, max_attempts: int):
+            self.retry_attempts += 1
+            if self.retry_attempts < max_attempts:
+                raise OSError(
+                    "When creating key x in bucket y: AWS Error SLOW_DOWN during "
+                    "PutObject operation: Please reduce your request rate."
+                )
+            return 0
+
+    counter = Counter()
+    assert _open_file_with_retry("dummy", lambda: counter.foo(3)) == 0
+
+    original_max_attempts = OPEN_FILE_MAX_ATTEMPTS
+    ray.data.datasource.file_based_datasource.OPEN_FILE_MAX_ATTEMPTS = 3
+    counter = Counter()
+    with pytest.raises(OSError):
+        _open_file_with_retry("dummy", lambda: counter.foo(4))
+    ray.data.datasource.file_based_datasource.OPEN_FILE_MAX_ATTEMPTS = (
+        original_max_attempts
+    )
```

Comment:
nit, patch this attribute, so it can restore the original value even if the test fails. 
Example:
```
def test_resource_constrained_triggers_autoscaling(monkeypatch):
    RESOURCE_REQUEST_TIMEOUT = 5
    monkeypatch.setattr(
        ray.data._internal.execution.autoscaling_requester,
        "RESOURCE_REQUEST_TIMEOUT",
        RESOURCE_REQUEST_TIMEOUT,
    )
```

Reply:
Changed to `try: finally:` to restore for now. I feel that already simple enough.